### PR TITLE
Improve _apply(apply_type, ::Tuple, ::SimpleVector)

### PR DESCRIPTION
**Pull request by Keno**
This is another one of those patterns that doesn't generally
come up except in Cassette/Zygote code. There's two related
changes here:

1. Expand ininling's _apply rewrite to also handle constant
   svecs (under the restriction that they must be no longer than
   the splatting cutoff and the elements must be individually eligible
   for inlining into the IR)
2. Move the _apply rewrite before the special case inliner for builtins
   such that _apply(apply_type, ...) gets eliminated early.

### CODE REVIEW

**Comment by Keno** [comment]
Hmm, passes tests, but seems to be causing some miscompiles elsewhere. Looking into it.

**Comment by Keno** [comment]
Never mind, I did a stupid. Fixed.

**Comment by Keno** [comment]
(For those following along at home, this fixes compiling the code generated by Zygote for the `crossentropy` loss function to TPUs)

File: base/compiler/ssair/inlining.jl [hunk 1]
```diff
@@ -606,8 +606,12 @@ function rewrite_apply_exprargs!(ir::IRCode, idx::Int, argexprs::Vector{Any}, at
         # now push flattened types into new_atypes and getfield exprs into new_argexprs
         for j in 1:length(def_atypes)
             def_atype = def_atypes[j]
-            new_call = Expr(:call, Core.getfield, def, j)
-            new_argexpr = insert_node!(ir, idx, def_atype, new_call)
+            if isa(def_atype, Const) && is_inlineable_constant(def_atype.val)
+                new_argexpr = insert_node!(ir, idx, def_atype, quoted(def_atype.val))
```
Comment (vtjnash) [inline base/compiler/ssair/inlining.jl]: `Const` shouldn't need a node at all, right, just use the value as the argexpr?

File: test/compiler/compiler.jl [hunk 2]
```diff
@@ -2086,3 +2086,13 @@ let ci = code_typed(bar_inlining_apply, Tuple{})[1].first
...
+
+# Test that inlining can inline _applys of builtins/_applys on SimpleVectors
+function foo_apply_apply_type_svec()
+    A = (Tuple, Float32)
+    B = Tuple{Float32, Float32}
+    Core.apply_type(A..., B.types...)
+end
+let ci = code_typed(foo_apply_apply_type_svec, Tuple{})[1].first
+    @test length(ci.code) == 1
+end
```
Comment (vtjnash) [inline test/compiler/compiler.jl]: can we make a better test than this? this one seems brittle under various compiler flags (notable, the code-coverage bot, although `code_typed` itself likes to mislead about this particular form of result also)—although perhaps this is really the best we can do, since we need to check for a particular optimization result?

perhaps we could at least improve the test to be more explicit: `ci.code == Any[ Expr(:return, Const(NTuple{3, Float32})) ]` (e.g. the expression that `code_typed` may or may not decide to replace the optimization result with)

**Review by vtjnash — COMMENTED** [review]
lgtm

File: base/compiler/ssair/inlining.jl (hunk 1 above)
Comment (Keno) [reply base/compiler/ssair/inlining.jl]: yes, I suspect that's true. I'll try it.

**Comment by jrevels** [comment]
Maybe not strictly relevant to this PR, but also ref https://github.com/jrevels/Cassette.jl/blob/cfa248a99209d17e8d1997dd887a698c4bc079b9/src/context.jl#L444 (in case you run into needing a similar specialization hack for your TPU purposes)

**Comment by Keno** [comment]
Somehow I though this had gotten merged and then got a rude awakening when this caused me bad performance when evaluating #31253. Rebased and addressed review comments.
